crates/shared/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -86,166 +100,365 @@ impl TryFrom<PoolData> for PoolInfo {
+            UniswapV3PoolEventFetcher(web3.clone()),
+            RecentEventsCache::default(),
+            Some(init_block),
+        ));
+
+        Ok(Self { checkpoint, events })
+    }
+
+    /// Moves the checkpoint to the block `latest_block - MAX_REORG_BLOCK_COUNT`
+    async fn move_checkpoint_to_future(&self) -> Result<()> {
+        let last_event_block = self.events.lock().await.store().last_event_block().await?;
+        let old_checkpoint_block = self
+            .checkpoint
+            .pools_checkpoint
+            .lock()
+            .unwrap()
+            .block_number;
+        let new_checkpoint_block = std::cmp::max(
+            last_event_block.saturating_sub(MAX_REORG_BLOCK_COUNT),
+            old_checkpoint_block,
+        );
+
+        if new_checkpoint_block > old_checkpoint_block {
+            let block_range =
+                RangeInclusive::try_new(old_checkpoint_block + 1, new_checkpoint_block)?;
+            let events = self.events.lock().await.store().get_events(block_range);
+            let mut checkpoint = self.checkpoint.pools_checkpoint.lock().unwrap();
+            append_events(&mut checkpoint.pools, events);
+            checkpoint.block_number = new_checkpoint_block;
+        }
+        Ok(())
+    }
 }
 
 #[async_trait::async_trait]
 impl PoolFetching for UniswapV3PoolFetcher {
-    async fn fetch(&self, token_pairs: &HashSet<TokenPair>) -> Result<Vec<PoolInfo>> {
-        tracing::debug!("token_pairs {:?}", token_pairs);
-        let (mut cached_pools, outdated_pools) = self.get_cached_pools(token_pairs);
-        tracing::debug!(
-            "cached pools: {:?}, outdated pools: {:?}",
-            cached_pools,
-            outdated_pools
-        );
+    async fn fetch(
+        &self,
+        token_pairs: &HashSet<TokenPair>,
+        at_block: Block,
+    ) -> Result<Vec<PoolInfo>> {
+        let block_number = match at_block {
+            Block::Recent => self
+                .events
+                .lock()
+                .await
+                .store()
+                .last_event_block()
+                .await
+                .unwrap_or_default(),
+            Block::Number(number) => number,
+        };
 
-        if !outdated_pools.is_empty() {
-            let updated_pools = self.get_pools_and_update_cache(&outdated_pools).await?;
-            tracing::debug!("updated pools: {:?}", updated_pools);
-            cached_pools.extend(updated_pools);
+        // sanity check if events are up to date
+        let last_handled_block = self
+            .events
+            .lock()
+            .await
+            .last_handled_block()
+            .unwrap_or_default()
+            .0;
+        if block_number > last_handled_block {
+            tracing::debug!(
+                "can't get liquidity for block {} since the last handled block is {}",
+                block_number,
+                last_handled_block
+            );
+            return Ok(Default::default());
```

Comment:
In this case, shouldn't we update to the block. The idea is that we use `Recent` if we want whatever is latest for the UniV3 fetcher and when we specify a block number we want the pools at that block (either from cache, or fetched).

Reply:
Theoretically this case should never happen because maintenance should be triggered on each new block and events should be updated immediately, so calling the event update explicitly at this point in the code would be redundant I assume.  But keeping this log is for safety or if the event update takes too long (in any case triggering it again here I don't see a purpose)

Reply:
Aaand I see this is true if `current_block_stream` polls really frequently (currently on each 5 sec). Will need to think more about this.

Reply:
Ended up calling `run_maintenance()` for events only, this should do the trick.